Add accessibility shortcoming select-input to Sports search

nikomakela opened the pull request:
LIIKUNTA-567

1. Fix noOutline for select-component.
2. Find a way how to use HDS-select as a dropdown-search-filter-field. There were some issues with the value handling, when the value was cleared.
3. Add a accessibilityFilter
4. Connect dropdown to combined search form values

### Value selection

When selecting: 

<img width="821" alt="image" src="https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/389204/27b6c873-e47b-4ec2-a539-79ab1626658a">

After the search is triggered:

<img width="822" alt="image" src="https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/389204/98bd3f98-b80d-4422-9869-98de1a73409d">


### List view

<img width="872" alt="image" src="https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/389204/ecb543bf-1ee3-4733-84d0-e25958246880">

<img width="572" alt="image" src="https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/389204/dcb0857e-b977-402d-bf87-91729d6bf83b">

<img width="451" alt="image" src="https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/389204/f42f8da1-81d2-4a08-9909-f59d56e469ca">

### Map view

<img width="761" alt="image" src="https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/389204/e53aacd4-45f1-4459-b94c-9955c72ee95b">

<img width="530" alt="image" src="https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/389204/75a46c9f-4d9c-42f0-9d3b-b51a7877ae80">


<img width="432" alt="image" src="https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/389204/268d696b-7ce2-4a46-9f52-7d444c4ba407">

nikomakela committed:
refactor(search): move useFormValues hook to a new file

LIIKUNTA-567

nikomakela committed:
style: fix noOutline for select-component

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://tapahtumat-pr518.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr518.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __success__ for https://liikunta-pr518.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
**EVENTS GRAPHQL PROXY is deployed to platta: https://events-graphql-proxy-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**VENUE GRAPHQL PROXY is deployed to platta: https://venue-graphql-proxy-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**VENUE GRAPHQL PROXY is deployed to platta: https://venue-graphql-proxy-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://liikunta-pr518.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __success__ for https://tapahtumat-pr518.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr518.dev.hel.ninja :laughing::tada::tada::tada:

nikomakela committed:
feat(search): add order venues by accessibility -option when filtering with accessibility

LIIKUNTA-568 LIIKUNTA-567

karisal-anders commented:
Instead of having an empty row that is selected how about making it explicitly "Ei liikkumisrajoitteita" / "No mobility impairments" / "Inga rörelsehinder" like in palvelukartta.hel.fi:
![image](https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/77663720/6e90970d-48c8-45c7-8fac-33153d6e3c55)
It would be clearer IMHO that way and also make liikunta.hel.fi's accessibility stuff more like palvelukartta's.

karisal-anders commented on packages/common-i18n/src/locales/en/search.json:
```diff
@@ -64,5 +66,14 @@
   "openStreetMapContributors": "contributors",
-  "mapAccessibilityStatement": "Map. Currently map information is only accessible visually."
+  "mapAccessibilityStatement": "Map. Currently map information is only accessible visually.",
+  "accessibilityShortcoming": {
+    "noAccessibilityProfile": "No accessibility shortcoming",
+    "hearing_aid": "Hearing aid",
+    "reduced_mobility": "Reduced mobility",
+    "rollator": "Rollator",
+    "stroller": "Stroller",
+    "visually_impaired": "Visually impaired",
+    "wheelchair": "Wheelchair"
+  }
```
IMHO it'd be good to use the same translations here as in https://palvelukartta.hel.fi/en/ as we are using their exact translations for in the Finnish version already.

karisal-anders commented on packages/common-i18n/src/locales/sv/search.json:
```diff
@@ -67,5 +69,14 @@
   "openStreetMapContributors": "bidragsgivare",
-  "mapAccessibilityStatement": "Karta, För tillfället är kartans uppgifter tillgängliga endast visuellt."
+  "mapAccessibilityStatement": "Karta, För tillfället är kartans uppgifter tillgängliga endast visuellt.",
+  "accessibilityShortcoming": {
+    "noAccessibilityProfile": "Inga tillgänglighetsbegränsningar",
+    "hearing_aid": "Jag användar hörapparat",
+    "reduced_mobility": "Jag har rörelsehinder",
+    "rollator": "Jag användar rollator",
+    "stroller": "Jag går med barnvagn",
+    "visually_impaired": "Jag är synskadad",
+    "wheelchair": "Jag användar rullstol"
+  }
```
typo: användar → använder

karisal-anders commented on packages/components/src/components/select/__tests__/SearchSelect.test.tsx:
```diff
@@ -0,0 +1,90 @@
+import { AccessibilityProfile } from '../../../types';
+
+import SearchSelect from '../SearchSelect';
+// eslint-disable-next-line import/no-unresolved
```
Unneeded eslint-disable-next-line

karisal-anders commented on apps/sports-helsinki/src/domain/search/combinedSearch/hooks/useFormValues.tsx:
```diff
@@ -0,0 +1,80 @@
+    string[]
+  >(formValues.targetGroups);
+  const [targetGroupInput, setTargetGroupInput] = React.useState('');
+  const targetGroups = getTargetGroupOptions(t).sort(
+    sortExtendedCategoryOptions
+  );
```
```suggestion
  const targetGroups = getTargetGroupOptions(t);
```

To not revert changes made with PR #520

karisal-anders commented on apps/sports-helsinki/src/domain/search/combinedSearch/search.module.scss:
```diff
@@ -98,26 +98,6 @@ $buttonWidth: 180px;
   }
 }
 
-.autoSuggestRowWithTargetGroupFilter {
-  @extend .autoSuggestRowBase;
-
-  & > :first-child {
-    @include respond-above(s) {
-      grid-column: 1 / 3;
-    }
-  }
-}
-
-.autoSuggestRowWithoutTargetGroupFilter {
-  @extend .autoSuggestRowBase;
-
-  & > :first-child {
-    @include respond-above(m) {
-      grid-column: 1 / 4;
-    }
-  }
-}
-
```
This leaves the NEXT_PUBLIC_SHOW_TARGET_GROUP_FILTER feature flag functionality in a broken state i.e. you can't really turn off the target group filter anymore without breaking the user interface. But as the target group filter feature flag probably isn't needed anymore you could just remove it altogether IMHO.

nikomakela replied:
I did test that feature flag and it should work. Did you test it? I changed the CSS a bit.

karisal-anders replied:
Seems to work ok in the less wide views but in the wider views it leaves an empty gap in the UI:

![image](https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/77663720/e40b4b8f-c2f4-4f5d-a96e-3fd147ccbaa2)

![image](https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/77663720/a92b07be-95f5-4237-b4cf-bcd60bb2b902)

![image](https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/77663720/a1e55d97-a50b-45ab-8ab3-ccbb170ffb0a)

![image](https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/77663720/ec9667b6-b9e7-4ac6-a357-1129873276bd)

nikomakela replied:
I think that I'll leave it as it is. There is that 1 column white space, but it should not matter since the taget group becomes always active field and the feature flag will be removed.

karisal-anders commented on apps/sports-helsinki/src/domain/search/eventSearch/utils.tsx:
```diff
@@ -91,6 +97,19 @@ export const getTargetGroupOptions = (
     getCategoryOptions(targetGroup, TARGET_GROUP_DATA[targetGroup], t)
   );
 
+export const accessibilityProfileOptionValues = [
+  ...Object.values(AccessibilityProfile),
+] as const;
+
+export const getAccessibilityShortcomingOptions = (
+  t: TFunction,
+  accessibilities: typeof accessibilityProfileOptionValues = accessibilityProfileOptionValues
```
Why the need for doing this differently than in getSportsCategoryOptions and getTargetGroupOptions? Not to duplicate values? BTW typeof accessibilityProfileOptionValues is readonly AccessibilityProfile[] because of the use of Object.values, if you want to have the union type of the enum's values you could make a type e.g.
```
type AccessibilityProfileValues = `${AccessibilityProfile}`
```

The shown accessibilities in the UI sketches do not include the hearing aid and the visually impaired options, for myself it is ok to leave them as they are now but I think it would be good to discuss the shown accessibilities with the designer/product owner as from earlier discussions I had I got the impression that the UI sketches in the ticket LIIKUNTA-27 show the ones wanted here.

karisal-anders reviewed (changes requested):
Left some comments, please have a look.

Request changes:
- https://github.com/City-of-Helsinki/events-helsinki-monorepo/pull/518/files#r1363410880 needs to be taken into account in order not to revert changes made with PR #520.

nikomakela committed:
feat(search): add accessibility profile as an option to venue search params

LIIKUNTA-568 LIIKUNTA-569 LIIKUNTA-570

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**EVENTS GRAPHQL PROXY is deployed to platta: https://events-graphql-proxy-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __failed__ for https://harrastukset-pr518.dev.hel.ninja :crying_cat_face::anger::boom::boom:

terovirtanen commented:
**VENUE GRAPHQL PROXY is deployed to platta: https://venue-graphql-proxy-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __failed__ for https://liikunta-pr518.dev.hel.ninja :crying_cat_face::anger::boom::boom:

nikomakela committed:
feat(search): sync the venue search ordering with the accessibility profile

LIIKUNTA-568 LIIKUNTA-569 LIIKUNTA-570.

When the accessibility profile is selected in the search context,
the venue ordering should always be a orderByAccessibilityProfile
with the current accessibility profile.

This can be achieved by controlling the venueOrderBy-parameter of the search form.
Every time when the accessibility profile is set and it does not match the venueOrderBy,
the venueOrderBy is set and the URL is refreshed.

Also, when the accessibility profile is set and the ordering should be forced
to accessibility profile, the ordering selector is disabled.

karisal-anders commented on apps/sports-helsinki/src/domain/search/combinedSearch/search.module.scss:
```diff
@@ -98,26 +98,6 @@ $buttonWidth: 180px;
 }
 
-.autoSuggestRowWithTargetGroupFilter {
-  @extend .autoSuggestRowBase;
```
It would be good to remove `.autoSuggestRowBase` also completely as it is now left unused by your changes.

karisal-anders commented on packages/common-i18n/src/locales/fi/search.json:
```diff
@@ -60,9 +60,20 @@
-    }
+    },
+    "labelAccessibilityProfile": "Esteettömyys",
+    "ariaLabelClearAccessibilityProfile": "Poista esteettömyys valinta"
```
`esteettömyys valinta` is a compound word i.e. `esteettömyysvalinta` (Also thought that could it be `esteettömyys-valinta`, but kielitoimisto's guides [#1](http://www.kielitoimistonohjepankki.fi/ohje/80) & [#2](http://www.kielitoimistonohjepankki.fi/ohje/130) would suggest that would not be so).

karisal-anders commented on packages/components/src/components/select/SearchSelect.tsx:
```diff
@@ -0,0 +1,39 @@
+      'The SearchSelect does not support multiselect feature. Use the <MultiSelectDropdown/> instead.'
+    );
+  }
+  if (Array.isArray(value)) {
+    throw new Error(
+      'The value must be a singleton! The SearchSelect does not support an array as a value feature. Maybe you should use the <MultiSelectDropdown/> instead.'
+    );
+  }
+  if (Array.isArray(defaultValue)) {
+    throw new Error(
+      'The defaultValue must be a singleton!The SearchSelect does not support an array as a value feature. Maybe you should use the <MultiSelectDropdown/> instead.'
+    );
+  }
+  // The value can be an array of options, a single option, null or undefined.
+  // The option value can also be empty!
+  const hasValue = Array.isArray(value) ? value.length : !!value?.value;
```
```suggestion
  if (Array.isArray(value) || Array.isArray(defaultValue)) {
    throw new Error(
      'Values must be singletons! The SearchSelect does not support an array as a value feature. Maybe you should use the <MultiSelectDropdown/> instead.'
    );
  }
  // The value can be a single option, null or undefined.
  // The option value can also be empty!
  const hasValue = !!value?.value;
```

A suggestion to combine long texts together for brevity and remove the hasValue case that never happens.

nikomakela committed:
refactor: removed the showTargetGroupFilter feature-flag

LIIKUNTA-567

nikomakela committed:
feat(search): add accessibility dropdown

LIIKUNTA-567.

Connect accessibility dropdown to form values

The HDS-select field value handling is bugged:
When the value is cleared from outside, for example from the clear filters -button,
the selected value indicator stays active in the HDS-select component.

Sort accessibility shortcoming options alphabetically.

The accessibility profile translations matches with the service map ones.

Use a limited list of accessibility profiles.

nikomakela committed:
refactor: improve the isAccessibilityProfile-typeguard

LIIKUNTA-568

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr518.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __success__ for https://tapahtumat-pr518.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
**EVENTS GRAPHQL PROXY is deployed to platta: https://events-graphql-proxy-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://liikunta-pr518.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
**VENUE GRAPHQL PROXY is deployed to platta: https://venue-graphql-proxy-pr518.dev.hel.ninja** :rocket::rocket::rocket:

nikomakela committed:
feat(search): venueOrderBy can be changed after accessibilityProfile selection

LIIKUNTA-568.
It was mentioned (not written) in the refinement meeting, that it would be nice,
if the `venueOrderBy`would be set to the value of `accessibilityProfile`, when the profile
is changed and one is set, but it could be changed to something else with a reselection.
This means that the `venueOrderBy` cannot be disabled anymore.

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr518.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __success__ for https://tapahtumat-pr518.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**VENUE GRAPHQL PROXY is deployed to platta: https://venue-graphql-proxy-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**EVENTS GRAPHQL PROXY is deployed to platta: https://events-graphql-proxy-pr518.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://liikunta-pr518.dev.hel.ninja :laughing::tada::tada::tada:
